lib/vrl/stdlib/src/parse_csv.rs

```diff
@@ -0,0 +1,132 @@
+use bytes::Buf;
+use csv::ReaderBuilder;
+use vrl::prelude::*;
+
+#[derive(Clone, Copy, Debug)]
+pub struct ParseCsv;
+
+impl Function for ParseCsv {
+    fn identifier(&self) -> &'static str {
+        "parse_csv"
+    }
+
+    fn examples(&self) -> &'static [Example] {
+        &[Example {
+            title: "parse a single CSV formatted row",
+            source: r#"parse_csv!(s'foo,bar,"foo "", bar"')"#,
+            result: Ok(r#"["foo", "bar", "foo \", bar"]"#),
+        }]
+    }
+
+    fn compile(&self, mut arguments: ArgumentList) -> Compiled {
+        let value = arguments.required("value");
+        let delimiter = arguments.optional("delimiter").unwrap_or(expr!(","));
+        Ok(Box::new(ParseCsvFn { value, delimiter }))
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[
+            Parameter {
+                keyword: "value",
+                kind: kind::BYTES,
+                required: true,
+            },
+            Parameter {
+                keyword: "delimiter",
+                kind: kind::BYTES,
+                required: false,
+            },
+        ]
+    }
+}
+
+#[derive(Debug, Clone)]
+struct ParseCsvFn {
+    value: Box<dyn Expression>,
+    delimiter: Box<dyn Expression>,
+}
+
+impl Expression for ParseCsvFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let csv_string = self.value.resolve(ctx)?.try_bytes()?;
+        let delimiter = self.delimiter.resolve(ctx)?.try_bytes()?;
+        if delimiter.len() != 1 {
+            return Err("delimiter must be a single character".into());
+        }
+        let delimiter = delimiter[0];
+
+        let reader = ReaderBuilder::new()
+            .has_headers(false)
+            .delimiter(delimiter)
+            .from_reader(csv_string.bytes());
+
+        let parsed_record = reader.into_records().next();
+        match parsed_record {
+            Some(Ok(record)) => {
+                let result = record.iter().map(|l| l.into()).collect();
+                Ok(Value::Array(result))
+            }
+            Some(Err(_)) => Err("failed to parse csv string".into()),
```

Comment:
```suggestion
            Some(Err(err)) => Err(format!("failed to parse csv string: {}", err).into()),
```

Reply:
I refactored it as @JeanMertz  suggested, though I couldn't find a good test case - seems like we shouldn't get any errors here - before I could fail it with invalid utf8, but not anymore given that now we use byte records.